From 6173ef8682540bb024a8f26cb6c2ba93788d23b2 Mon Sep 17 00:00:00 2001
From: Patrick Delaunay <patrick.delaunay@foss.st.com>
Date: Tue, 15 Jul 2025 11:43:11 +0200
Subject: [PATCH 898/954] feat(stm32mp2): fix support for CA35SS voltage
 scaling

In STM32MP2 family the CA35 DVFS operation required a specific sequence
to change the VDDCPU and Cortex-A35 clock and to control the speed of
memories (RM) inside Cortex-A35.

This patch change the sequence implemented in the associated SVC:
- set PLL1 in bypass
- OPP request with RM configuration to nominal if needed
- Change Voltage (if needed)
- Switch PLL1 to target frequency
- OPP request with RM configuration to overdrive if needed

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 39b64dc40b844f9a1c7c8c2169e50b66dafd8869 feat(stm32mp2): fix support for CA35SS voltage scaling]
Signed-off-by: Patrick Delaunay <patrick.delaunay@foss.st.com>
Change-Id: Id98115e5b37d74c1f85b6db8b73266483469cfcf
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/475535
Reviewed-by: Maxime MERE <maxime.mere@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 plat/st/stm32mp2/services/ca35ss_clk_svc.c | 73 ++++++++++++++--------
 1 file changed, 48 insertions(+), 25 deletions(-)

diff --git a/plat/st/stm32mp2/services/ca35ss_clk_svc.c b/plat/st/stm32mp2/services/ca35ss_clk_svc.c
index ce6b20986e..0718f51cf0 100644
--- a/plat/st/stm32mp2/services/ca35ss_clk_svc.c
+++ b/plat/st/stm32mp2/services/ca35ss_clk_svc.c
@@ -19,6 +19,7 @@
 
 #include "ca35ss_clk_svc.h"
 #include "scmi_private.h"
+#include <stm32mp2_private.h>
 #include <stm32mp2_smc.h>
 #include <stm32mp_common.h>
 #include <stm32mp_svc_setup.h>
@@ -40,6 +41,9 @@
 #define VOLTD_SCMI_STPMIC2_BUCK		VOLTD_SCMI_STPMIC2_BUCK3
 #endif
 
+/* Nominal rate for CA33SS in STM32MP2, to detect overdrive freq */
+#define NOMINAL_RATE			1200000
+
 enum stm32_ca35ss_clk_svc_function {
 	CA35SS_CLK_SVC_NO_FUNCTION,
 	CA35SS_CLK_SVC_SET_RATE,
@@ -61,7 +65,6 @@ struct ca35ss_state
 {
 	int32_t state;
 	int32_t target_opp_idx;
-	bool pll1_done;
 };
 
 struct ca35ss_opp
@@ -100,6 +103,7 @@ static struct ca35ss_state state;
  *    |               | --------------> |                 |       GET_VOLT_POLL
  *    |               |       volt      |                 |       GET_VOLT_POLL
  *    |               | < - - - - - - - |                 |       GET_VOLT_POLL
+ *    |               |  << set CA35 clock in bypass >>
  *    |               |     set_volt    |                 |       SET_VOLT_POLL
  *    |               | --------------> |                 |       SET_VOLT_POLL
  *    |               |              Doorbell             |       SET_VOLT_POLL
@@ -109,7 +113,7 @@ static struct ca35ss_state state;
  *                      [POLLING UNTIL CHANNEL IS FREE]
  *    |    poll()     |                 |                 |       SET_VOLT_POLL
  *    | ------------> |                 |                 |       SET_VOLT_POLL
- *    |      OK       |                 |                 |       BASE
+ *    |      OK       | << set PLL1 with target frequency >>
  *    | < - - - - - - |                 |                 |       BASE
  *
  * Hence, from Linux the view is as simple as a set_rate() followed by a status
@@ -121,7 +125,6 @@ static void ca35ss_set_state_base(void)
 {
 	state.state = CA35SS_CLK_SVC_STATE_BASE;
 	state.target_opp_idx = -1;
-	state.pll1_done = false;
 }
 
 /* Device tree parsing functions */
@@ -423,6 +426,16 @@ uint32_t ca35ss_clk_svc_setup(void)
 		return STM32_SMC_FAILED;
 	}
 
+	/* Verify supported setting for PLL1 */
+	for (opp_idx = 0; opp_idx < opp.opp_nb; opp_idx++) {
+		err = stm32mp2_pll1_check_rate(opp.hz[opp_idx]);
+		if (err != 0) {
+			ERROR("%s: PLL1 configuration for %lu Hz not found (%d)\n",
+			      __func__, opp.hz[opp_idx], err);
+			return STM32_SMC_FAILED;
+		}
+	}
+
 	/* Verify SCMI availlability */
 	timeout_ref = timeout_init_us(TIMEOUT_10MS_IN_US);
 	while (scmi_channel_busy())
@@ -519,12 +532,30 @@ static uint32_t ca35ss_clk_svc_handler_set_rate(uint64_t target_rate)
 	return status;
 }
 
+static uint32_t ca35ss_clk_svc_pll1_set(uint64_t target_rate)
+{
+	uint32_t status;
+
+	/* Change OPP/RM(overdrive) when moving OPP Nominal to Overdrive */
+	if (target_rate > NOMINAL_RATE) {
+		stm32mp_ca35_configure_opp(true);
+	}
+
+	/* Change PLL1 rate  */
+	status = stm32mp2_pll1_set_rate(target_rate);
+
+	/* Back to base state */
+	ca35ss_set_state_base();
+
+	return status;
+}
+
 static uint32_t ca35ss_clk_svc_handler_poll_get_volt()
 {
 	uint32_t status;
-	int32_t err;
 	int32_t current_uvolt;
 	int32_t opp_idx = state.target_opp_idx;
+	uint64_t target_rate = opp.hz[state.target_opp_idx];
 
 	/* Poll the status of the operation of getting of current vddcpu */
 	status = vddcpu_get_uvolt_poll(&current_uvolt);
@@ -536,21 +567,20 @@ static uint32_t ca35ss_clk_svc_handler_poll_get_volt()
 		return status;
 	}
 
-	/* Next step */
-	if (current_uvolt >= (int32_t)opp.uvolt[opp_idx]) {
-		/* If dicreasing voltage, then frequency scaling comes first */
-		err = stm32mp2_pll1_set_rate(opp.hz[opp_idx]);
-		if (err != 0) {
-			/* Back to base state */
-			ca35ss_set_state_base();
-			return STM32_SMC_FAILED;
-		}
+	/* Switch to bypass, PLL1 no more used */
+	stm32mp2_pll1_set_rate(0U);
 
-		/* pll1 has been scaled */
-		state.pll1_done = true;
+	/* Change OPP/RM(nominal) when moving OPP Overdrive to Nominal */
+	if (target_rate <= NOMINAL_RATE) {
+		stm32mp_ca35_configure_opp(false);
 	}
 
-	/* Initiate the operation of scaling vddcpu */
+	/* Check update voltage need*/
+	if (current_uvolt == (int32_t)opp.uvolt[opp_idx]) {
+		return ca35ss_clk_svc_pll1_set(target_rate);
+	}
+
+	/* Next step :  Update voltage */
 	status = vddcpu_set_uvolt(opp_idx);
 	if (status != STM32_SMC_ON_GOING) {
 		/* Back to base state */
@@ -575,15 +605,8 @@ static uint32_t ca35ss_clk_svc_handler_poll_set_volt(void)
 		return STM32_SMC_ON_GOING;
 	}
 
-	if ((status == STM32_SMC_OK) && (!state.pll1_done)) {
-		/* Change PLL1 rate if not done yet */
-		status = stm32mp2_pll1_set_rate(target_rate);
-	}
-
-	/* Back to base state */
-	ca35ss_set_state_base();
-
-	return status;
+	/* Change PLL1 rate  */
+	return ca35ss_clk_svc_pll1_set(target_rate);
 }
 
 static uint32_t ca35ss_clk_svc_handler_set_rate_status(void)
-- 
2.53.0

